### ISSUE TITLE
Mixture Density Networks | 大トロ

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#1379]
Mixture Density Networks | 大トロ<br>
<br>
https://ift.tt/1P3CA8d<br>
<br>
<br>
<br>

<div>
<section><p><a href="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/Screen-Shot-2015-06-14-at-10.52.20-AM.png"><img src="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/Screen-Shot-2015-06-14-at-10.52.20-AM-1024x590.png" alt="Screen Shot 2015-06-14 at 10.52.20 AM"></a></p>

<p>For the Javascript demo of Mixture Density Networks, here is the <a href="http://otoro.net/ml/mixture/mixture.html">link</a>.</p>

<p><em>Update: A more comprehensive write-up about MDNs implemented with TensorFlow <a href="http://blog.otoro.net/2015/11/24/mixture-density-networks-with-tensorflow/">here</a></em></p>

<p>While I was going through Grave’s <a href="http://www.cs.toronto.edu/~graves/handwriting.html">paper</a> on artificial handwriting generation, I noticed that his model is not setup to predict the next location of the pen, but trained to generate a probability distribution of what happens next to the pen, including whether then pen gets lifted up.  It eventually made sense to me, since if we train the model to predict the next location based on some mean squared error minimisation function the model would not get anywhere as the average of the next step may not contain much information.  The algorithm actually generated a gaussian mixture distribution used for predicting the next location of the pen, and also a simple bernoulli distribution for whether the pen stays on the writing pad or not.  It was a bit over my head as this concept is then combined with <a href="https://en.wikipedia.org/wiki/Long_short_term_memory">Long Short Term Memory</a> neural nets and conditional gates for handwriting synthesis.</p>

<p>After a few discussions with <a href="https://twitter.com/QuantixResearch">Quantix</a> who advised me to go through some useful literature, I became interested with an old paper about <a href="http://research.microsoft.com/en-us/um/people/cmbishop/downloads/Bishop-NCRG-94-004.ps">Mixture Density Networks</a> approach developed by <a href="https://en.wikipedia.org/wiki/Christopher_Bishop">Bishops</a> way back in the mid 1990’s, who now works for <a href="http://research.microsoft.com/en-us/um/people/cmbishop/">Microsoft Research</a>.  I thought that the paper was very well written and the author’s style was more about teaching the reader how something works rather than promoting their research.  I wish more papers nowadays were written in the same style.</p>

<p>Mixture density models are nothing new and they are used everywhere to estimate the real distribution of some data, typically by assuming that each datapoint has some probability to be associated to a certain gaussian distribution, as in the equations below:</p>

<p><a href="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/mixture_formula.png"><img src="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/mixture_formula.png" alt="mixture_formula"></a> <a href="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/normal.png"><img src="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/normal.png" alt="normal"></a></p>

<p>The mean square minimisation approach can be regarded as similar to the special case of the mixture gaussian model when m=1 (but not exactly the same).</p>

<p>This sort of model can be useful if combined with neural networks, where the outputs of the neural network are the <em>parameters</em> of the mixture model, rather than direct prediction of the data label.  So for each input, you would have a set of mean parameters, a set of standard deviation parameters, and a set of probabilities that the output point would fall into those gaussian distributions.</p>

<p>Bishop’s paper mentioned that often in machine learning problems, we have to solve <em>inverse problems</em>, and we are interested in learning about some state that get us the correct answer.  For example, given some medical symptoms, estimate a list of likely diseases that the patient may have.  Or, given a desired location of a robotic arm, figure out which angles should all the robotic joints rotate towards, so that the robotic arm can get to the desired location.  For those problems, the average of the possible answers is unlikely to be the correct answer, so training the algorithm using a mean square error method will not work.</p>

<p>The Mixture Density Networks paper outlines the method to estimate the conditional mixture density parameters given some input, by minimising the cross entropy of the distribution mentioned above.<a href="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/error.png"><img src="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/error.png" alt="error"></a></p>

<p>His paper included derivations on closed form gradient formulas making it possible to train the network efficiently using back propagation.  I attempted to implement his algorithm to tackle some toy problems.  As I wanted to make the algorithm work on a web browser, I used the graph objects from karpathy’s <a href="https://github.com/karpathy/recurrentjs">recurrent.js</a> library for the implementation.  Note: there is actually no recurrent networks here, but I just liked how things are done in recurrent.js compared to the older convnet.js library, and features such as auto differentiation is possible, and in the future I can extend this code to do much cooler stuff.  You can click on the graphs below to see the training of the network done interactively in your web browser.</p>

<p>As a toy problem, we generate some random data, say based off this formula, a sinusoid with a linear bias and some gaussian noise, and try to get our model to fit the output.</p>

<p><a href="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/data_formula.png"><img src="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/data_formula.png" alt="data_formula"></a></p>

<p>The data can be easily fitted with a 2-layer network and 5 hidden tanh units (see demo <a href="http://otoro.net/ml/mixture/index.html">here</a>):</p>

<div id="attachment_575" class="wp-caption alignnone">
  <a href="http://otoro.net/ml/mixture/index.html"><img class="wp-image-575 size-large" src="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/square_error_data-1024x1022.png" alt="square_error_data"></a>

  <p class="wp-caption-text">
    data fitting by minimising the square error
  </p>
</div>

<p> </p>

<p>However, if we use the inverse of the data above, ie, the y’s become the inputs, and the x’s then become the desired outputs, we would have problems using the minimum square error approach, since for each input, there are multiple outputs that would work, and the network would likely predict some sort of average of the correct values.  If we run the same algorithm on the inverse data, this is what we get (see demo <a href="http://otoro.net/ml/mixture/inverse.html">here</a>):</p>

<div id="attachment_577" class="wp-caption alignnone">
  <a href="http://otoro.net/ml/mixture/inverse.html"><img class="wp-image-577 size-large" src="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/square_error_inverse-1014x1024.png" alt="square_error_inverse"></a>

  <p class="wp-caption-text">
    minimum square error data fitting of inverse data
  </p>
</div>

<p> </p>

<p>Bishop’s algorithm would have the network output a set of parameters for the gaussian mixture model, given each x value.  My implementation would sample ten values from that distribution, and plot those samples on the y-axis, to get a feel of the multi-modal distribution.  I have also played around with the number of mixtures needed to get it to work, and it seems that any value greater than 5 would do okay for this simple set of data.  Below is how the mixture density network model fits the inverse data (see demo <a href="http://otoro.net/ml/mixture/mixture.html">here</a>):</p>

<div id="attachment_578" class="wp-caption alignnone">
  <a href="http://otoro.net/ml/mixture/mixture.html"><img src="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/mixture_distribution_inverse-1007x1024.png" alt="mixture density network predicting a distribution for the output, rather than a single value, for each input"></a>

  <p class="wp-caption-text">
    mixture density network predicting a distribution for the output, rather than a single value, for each input
  </p>
</div>

<p> </p>

<p>We see that there are still some noise around the edges, and the convergence time is slower than the simple minimum square error approach, but I’m satisfied that it all actually works!</p>

<p>Just as a check, the mixture model would also get the original data set correct, although it is a bit slower, and would obviously be overkill.</p>

<p><a href="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/mixture_distribution_data.png"><img src="http://blog.otoro.net/wp-content/uploads/sites/2/2015/06/mixture_distribution_data-1024x1024.png" alt="mixture_distribution_data"></a></p>
</section>

    
  
  </div><br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via blog.otoro.net http://blog.otoro.net<br>
<br>
October 31, 2018 at 05:24PM
